php/class-settings.php

```diff
@@ -1,270 +1,597 @@
+				$param['setting']                    = $this->add( $slug, $default, $param );
+			}
+		}
+
+		return $params;
+	}
+
+	/**
+	 * Magic method to get a chainable setting.
+	 *
+	 * @param string $name The name of the setting to get dynamically.
+	 *
+	 * @return Setting|null
 	 */
-	const SETTINGS_DATA = '_settings_version';
+	public function __get( $name ) {
+		$setting = null;
+		if ( isset( $this->settings[ $name ] ) ) {
+			$setting = $this->settings[ $name ];
+		}
+		if ( ! $setting ) {
+			$setting = $this->find_setting( $name );
+		}
+
+		return $setting;
+	}
 
 	/**
-	 * Initiate the settings object.
+	 * Remove a setting.
+	 *
+	 * @param string $slug The setting to remove.
+	 *
+	 * @return bool
 	 */
-	protected function __construct() {
-		add_action( 'admin_init', array( $this, 'register_wordpress_settings' ) );
-		add_action( 'admin_menu', array( $this, 'build_menus' ) );
+	public function delete( $slug ) {
+		$this->remove_param( self::META_KEYS['data'] . $this->separator . $slug );
+
+		return $this->storage->delete( $slug );
 	}
 
 	/**
-	 * Check settings version to allow settings to update or upgrade.
+	 * Init the settings.
+	 */
+	protected function init() {
+		$storage       = $this->get_param( 'storage' );
+		$this->storage = new $storage( $this->slug );
+	}
+
+	/**
+	 * Register a settings storage point.
 	 *
-	 * @param string $slug The slug for the settings set to check.
+	 * @param string $slug The key (option-name) to register the storage as.
 	 */
-	protected static function check_version( $slug ) {
-		$key              = '_' . $slug . self::SETTINGS_DATA;
-		$settings_version = get_option( $key, 2.4 );
-		$plugin_version   = get_plugin_instance()->version;
-		if ( version_compare( $settings_version, $plugin_version, '<' ) ) {
-			// Allow for updating.
-			do_action( "{$slug}_settings_upgrade", $settings_version, $plugin_version );
-			// Update version.
-			update_option( $key, $plugin_version );
+	protected function register_storage( $slug ) {
+		// Get the root key.
+		$slug = explode( $this->separator, $slug, 2 )[0];
+		if ( ! isset( $this->storage_keys[ $slug ] ) ) {
+			$data     = $this->storage->get( $slug );
+			$defaults = $this->get_param( self::META_KEYS['data'] . $this->separator . $slug, null );
+			if ( ! empty( $data ) ) {
+				if ( ! empty( $defaults ) && is_array( $data ) ) {
+					$data = wp_parse_args( $data, (array) $defaults );
+				}
+				$this->set_param( self::META_KEYS['data'] . $this->separator . $slug, $data );
+			}
+			$this->storage_keys[ $slug ] = $data;
 		}
 	}
 
 	/**
-	 * Register the page.
+	 * Get the storage key.
+	 *
+	 * @param string $slug The slug to get.
+	 * @param string $type The setting type.
+	 *
+	 * @return string
 	 */
-	public function build_menus() {
-		foreach ( $this->settings as $setting ) {
-			$this->register_admin( $setting );
+	protected function get_storage_key( $slug, $type = null ) {
+		if ( null == $type ) {
```

Comment:
@DavidCramer 

Isn't this a too loose comparison?

Reply:
===